src/include/detail/graph/vamana.h

```diff
@@ -0,0 +1,1019 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *
+ */
+
+#ifndef TDB_VAMANA_H
+#define TDB_VAMANA_H
+
+#include <cstddef>
+
+#include <functional>
+#include <queue>
+#include <unordered_set>
+
+#include "scoring.h"
+#include "stats.h"
+#include "utils/fixed_min_heap.h"
+#include "utils/print_types.h"
+
+#include "detail/graph/adj_list.h"
+#include "detail/graph/graph_utils.h"
+
+#include <tiledb/group_experimental.h>
+#include <tiledb/tiledb>
+
+namespace detail::graph {
+namespace {
+
+/**
+ * A tag to determine whether to return the top k results of the search or just
+ * the path taken.
+ */
+enum class SearchPath { path_and_search, path_only };
+
+/**
+ * An augmented distance functor that counts the number of distance
+ * invocations.
+ */
+struct counting_sum_of_squares_distance {
+  size_t num_comps_{0};
+  std::string msg_{""};
+
+  counting_sum_of_squares_distance() = default;
+  counting_sum_of_squares_distance(const std::string& msg)
+      : msg_(msg) {
+  }
+
+  template <class V, class U>
+  constexpr auto operator()(const V& a, const U& b) {
+    ++num_comps_;
+    return sum_of_squares(a, b);
+  }
+};
+}  // namespace
+
+/**
+ * @brief Truncated best-first search
+ * @tparam Distance The distance function used to compare vectors
+ * @param graph Graph to be searched
+ * @param source start node index
+ * @param query query node index
+ * @param k result size
+ * @param L search list size, L >= k
+ * @return result set ell containing k-approximate nearest neighbors to query
+ * and set vee containing all the visited nodes
+ *
+ * Per the DiskANN paper, the algorithm is as follows:
+ * 1. Initialize the result list with the source node and visited list with
+ *    empty
+ * 2. While the result list \ visited list is not empty
+ *    a. Find p* in the result list \ visited list with the smallest distance to
+ *       the query
+ *    b. update the result list with the out neighbors of p*
+ *    c. Add p* to the visited list d. If size of the result list > L, trim
+ *      the result list to keep L closest points to query
+ * 3. Copy the result list to the output
+ *
+ * This is essentially a best-first search with a fixed size priority queue
+ *
+ * @todo -- add a `SearchPath `template parameter to determine whether to
+ * return the top k results of the search or just the path taken.
+ * @todo -- remove printf debugging code
+ */
+template </* SearchPath SP, */ class Distance = sum_of_squares_distance>
+auto greedy_search(
+    auto&& graph,
+    auto&& db,
+    typename std::decay_t<decltype(graph)>::id_type source,
+    auto&& query,
+    size_t k_nn,
+    size_t L,
+    Distance&& distance = Distance{}) {
+  constexpr bool noisy = false;
+
+  // using feature_type = typename std::decay_t<decltype(graph)>::feature_type;
```

Comment:
(Minor) Remove commented out code? Going to stop reviewing for this specifically, but let's remove this throughout the PR.